### ISSUE TITLE
Unable to install with PIP

### ISSUE THREAD

**Issue by WhoCannotBeNamed** [theJollySin/mazelib#104]
Every single time i try to install mazelib I get this error:

----------------------------------------
      ERROR: Failed building wheel for numpy
      ERROR: Command errored out with exit status 1:
       command: /usr/local/bin/python -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-wheel-x7r47qyy/numpy/setup.py'"'"'; __file__='"'"'/tmp/pip-wheel-x7r47qyy/numpy/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' clean --all
           cwd: /tmp/pip-wheel-x7r47qyy/numpy
      Complete output (10 lines):
      Running from numpy source directory.
    
      `setup.py clean` is not supported, use one of the following instead:
    
        - `git clean -xdf` (cleans all files)
        - `git clean -Xdf` (cleans all versioned files, doesn't touch
                            files that aren't checked into the git repo)
    
      Add `--force` to your command to use it anyway if you must (unsupported).
    
      ----------------------------------------
      ERROR: Failed cleaning build dir for numpy
    ERROR: Failed to build one or more wheels
    Traceback (most recent call last):
      File "/usr/local/lib/python3.9/site-packages/setuptools/installer.py", line 128, in fetch_build_egg
        subprocess.check_call(cmd)
      File "/usr/local/lib/python3.9/subprocess.py", line 372, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['/usr/local/bin/python', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmp3z9slaw0', '--quiet', 'numpy<=1.16.4,>=1.13.1']' returned non-zero exit status 1.
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-5rn16r9y/mazelib/setup.py", line 60, in <module>
        setup(
      File "/usr/local/lib/python3.9/site-packages/setuptools/__init__.py", line 143, in setup
        _install_setup_requires(attrs)
      File "/usr/local/lib/python3.9/site-packages/setuptools/__init__.py", line 138, in _install_setup_requires
        dist.fetch_build_eggs(dist.setup_requires)
      File "/usr/local/lib/python3.9/site-packages/setuptools/dist.py", line 718, in fetch_build_eggs
        resolved_dists = pkg_resources.working_set.resolve(
      File "/usr/local/lib/python3.9/site-packages/pkg_resources/__init__.py", line 781, in resolve
        dist = best[req.key] = env.best_match(
      File "/usr/local/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1066, in best_match
        return self.obtain(req, installer)
      File "/usr/local/lib/python3.9/site-packages/pkg_resources/__init__.py", line 1078, in obtain
        return installer(requirement)
      File "/usr/local/lib/python3.9/site-packages/setuptools/dist.py", line 777, in fetch_build_egg
        return fetch_build_egg(self, req)
      File "/usr/local/lib/python3.9/site-packages/setuptools/installer.py", line 130, in fetch_build_egg
        raise DistutilsError(str(e))
    distutils.errors.DistutilsError: Command '['/usr/local/bin/python', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmp3z9slaw0', '--quiet', 'numpy<=1.16.4,>=1.13.1']' returned non-zero exit status 1.
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.

It is preceded with a big chunk of gcc debug information. Impossible to paste everything here.
I used the default docker Python image to assure it can be reproduced. Just get a shell inside the running image and run: 
```bash
> pip install Cython (flawless)
> pip install NumPy (flawless)
> pip install mazelib (error)
```

**Comment by theJollySin** [comment]
That's unfortunate.  I just tried and I was able to install from PIP without any problems.  So, probably we have different setups and I haven't tested this on yours.

Can you describe your system set up for me?  It seems possible that we have something different in our systems that is causing the problem.  Give me these kinds of details:

1. your operating system (and exact version)
2. what version of Python you are using  (e.g. 3.7.3)
3. Exactly how you are trying to install using PIP (like, the exact command line)
4. run this command line (if you can), and give me the screen dump: "pip freeze"

Numbers (1) and (2) are certainly important.

Slightly more advanced, but if you know VirtualEnv, can you try this from a clean Python v3.7 env and see what happens?  This is optional, for sure, but it can help weed out various confounding factors.

Sorry there's a problem, I have only tested this on one or two computers, obviously.  But I'm sure it won't be hard to sort out.

Thanks!

**Comment by theJollySin** [comment]
Do you have any information on this? What Operating System and version of Python are you using?

I have tried "pip install mazelib" at home and at work.  Python 3.6 and 3.7.  Operating Systems are all Linux though:  Debian 10.3, Ubuntu 16.04, and Ubuntu 18.04.

So far, I cannot reproduce your issue.  A little more information could help.

**Comment by WhoCannotBeNamed** [comment]
I'm sorry for the late response. I'm using Arch Linux as my Operating System with python version 3.8.1. 
I tested it on Docker as well to make sure it wasn't a problem with my specific configuration. It didn't work there (Docker python image uses python 3.9), you can test it by your own (https://hub.docker.com/_/python/).
If I use the python version 3.7 it works like a charm. The problem seems to appear with the new python versions.

**Comment by theJollySin** [comment]
Ah!

Please note in my setup.py I explicitly state that Python 3.8 and 3.9 are NOT supported. Supported version of Python are listed explicitly:

                 'Programming Language :: Python :: 3.3',
                 'Programming Language :: Python :: 3.4',
                 'Programming Language :: Python :: 3.5',
                 'Programming Language :: Python :: 3.6',
                 'Programming Language :: Python :: 3.7',

You can actually see this on the PyPI page for Mazelib, it states the supported versions of Python, and v3.8 and v3.9 are not supported (bottom left bar):

https://pypi.org/project/mazelib/

The only thing that surprises me here is that PyPI let you try to install on an unsupported language version.  That seems dangerous and bad.

I am closing this ticket.  BUT, I will open two more:

1. Find a way to block PyPI from trying to install in unsupported languages/versions.
2. See how much work it will be to support Python v3.8 and v3.9.

**Comment by chrisnorman7** [comment]
Hi,
Is there an issue for upgrading to Python 3.9? That's what I'm using, and I'd really like to use Mazelib.

Happy to look into porting myself obviously.

**Comment by theJollySin** [comment]
@chrisnorman7 @WhoCannotBeNamed 
Okay, the Cython team fixed whatever issue they were having and I can now support Python 3.8 and 3.9 in mazelib.

Just FYI.

**Comment by chrisnorman7** [comment]
Oh, so if I just update cython it should work?

I actually copied the source directory for mazelib, and it's all sememingly
working. Would be nice to have the c components mind, since I suspect
that'll speed things up a little.

Thanks for the info, it's very much appreciated.

Take care,

Chris Norman

**Comment by theJollySin** [comment]
@chrisnorman7  If you copied the source directory, you should be able to "python setup.py install" to build the Cython and get those C components yourself.

I don't like including the compiled C code in the repo, as it will change slightly based on the OS and Cython version you are using.  

Cheers!

**Comment by chrisnorman7** [comment]
OK, that works great, thank you.

Any news on when it'll make its way to pypi?

Thanks so much for your hard work.

Take care,

Chris Norman

**Comment by theJollySin** [comment]
@chrisnorman7  The latest version of mazelib is available on PyPI for Python 3.9:

https://pypi.org/project/mazelib/

**Comment by chrisnorman7** [comment]
Hi,
Unfortunately, it still doesn't work for me. The output is too long to paste, but I created a [Dropbox transfer](https://www.dropbox.com/t/EGzLaTHDQtKqKgRG).

Both numpy and cython are up-to-date.

Any help on what I could be doing wrong gratefully accepted.

**Comment by theJollySin** [comment]
Hello again @chrisnorman7 

The second line in that log file is concerning:

    Collecting mazelib
      Using cached mazelib-0.9.4.tar.gz (1.8 MB)

If you are trying to install in Python 3.9 you *have* to use `mazelib-0.9.10` (or newer).  That old `mazelib-0.9.4`... I'm not even sure if that supported Python 3.7.  So, if I understand what you're setup is like, I would say try this:

    pip uninstall -y mazelib
    pip uninstall -y mazelib
    pip uninstall -y mazelib
    pip install mazeliib==0.9.10

**Comment by chrisnorman7** [comment]
Hi,
Thanks for the heads up.

I'd have thought that using the -U switch would have gotten me the latest
version, which it might do, except that I don't seem to be able to install
the exact version either:

```bash
ERROR: Could not find a version that satisfies the requirement
mazelib==0.9.10
ERROR: No matching distribution found for mazelib==0.9.10
```

I actually get the same error (as well as Pip trying to install 0.9.4) on
my Raspberry Pi, also running Python 3.9.2.

Any other suggestions? :)

Cheers,

Take care,

Chris Norman

**Comment by theJollySin** [comment]
@chrisnorman7  And thus continues my love/hate relationship with PyPI.  I have removed the `0.9.10` release from PyPI entirely because it decided that it would only allow the latest version of mazelib to be installed up to Python v3.9.0, and would NOT allow v3.9.1.  So... that's idiotic.

I have fixed it, and we now have a mazelib version 0.9.11 which PyPI should stop complaining about.

My suggestion to you:

    pip uninstall -y mazelib
    pip uninstall -y mazelib
    pip uninstall -y mazelib
    pip install --no-cache-dir mazelib==0.9.11

I tried this with Python 3.9.2 and it worked fine (on Linux and MacOS) using PIP version 21.0.1.

Sorry for the hassle,
John
